src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -50,153 +51,137 @@ case class InferReadWriteAnnotation(t: String, tID: TransID)
 object InferReadWritePass extends Pass {
   def name = "Infer ReadWrite Ports"
 
-  def inferReadWrite(m: Module) = {
-    import WrappedExpression.we
-    val connects = HashMap[String, Expression]()
-    val repl = HashMap[String, Expression]()
-    val stmts = ArrayBuffer[Statement]()
-    val zero = we(UIntLiteral(0, IntWidth(1)))
-    val one = we(UIntLiteral(1, IntWidth(1)))
+  type Netlist = collection.mutable.HashMap[String, Expression]
+  type Statements = collection.mutable.ArrayBuffer[Statement]
+  type PortSet = collection.mutable.HashSet[String]
+
+  private implicit def toString(e: Expression) = e.serialize
 
-    // find all wire connections
-    def analyze(s: Statement): Unit = s match {
-      case s: Connect  =>
-        connects(s.loc.serialize) = s.expr
-      case s: PartialConnect =>
-        connects(s.loc.serialize) = s.expr
-      case s: DefNode =>
-        connects(s.name) = s.value
-      case s: Block =>
-        s.stmts foreach analyze
-      case _ =>
+  def getProductTerms(connects: Connects)(e: Expression): Seq[Expression] = e match {
+    // No ConstProp yet...
+    case Mux(cond, tval, fval, _) if weq(tval, one) && weq(fval, zero) =>
+      getProductTerms(connects)(cond)
+    // Visit each term of AND operation
+    case DoPrim(op, args, consts, tpe) if op == And =>
+      e +: (args flatMap getProductTerms(connects))
+    // Visit connected nodes to references
+    case _: WRef | _: WSubField | _: WSubIndex => connects get e match {
+      case None => Seq(e)
+      case Some(ex) => e +: getProductTerms(connects)(ex)
     }
+    // Otherwise just return itself
+    case _ => Seq(e)
+  }
 
-    def getProductTermsFromExp(e: Expression): Seq[Expression] =
-      e match {
-        // No ConstProp yet...
-        case Mux(cond, tval, fval, _) if we(tval) == one && we(fval) == zero =>
-          cond +: getProductTerms(cond.serialize)
-        // Visit each term of AND operation
-        case DoPrim(op, args, consts, tpe) if op == And =>
-          e +: (args flatMap getProductTermsFromExp)
-        // Visit connected nodes to references
-        case _: WRef | _: SubField | _: SubIndex | _: SubAccess =>
-          e +: getProductTerms(e.serialize)
-        // Otherwise just return itselt
-        case _ =>
-          List(e)
-      }
+  def checkComplement(a: Expression, b: Expression) = (a, b) match {
+    // b ?= Not(a)
+    case (_, DoPrim(Not, args, _, _)) => weq(args.head, a)
+    // a ?= Not(b)
+    case (DoPrim(Not, args, _, _), _) => weq(args.head, b)
+    // b ?= Eq(a, 0) or b ?= Eq(0, a)
+    case (_, DoPrim(Eq, args, _, _)) =>
+      weq(args(0), a) && weq(args(1), zero) ||
+      weq(args(1), a) && weq(args(0), zero)
+    // a ?= Eq(b, 0) or b ?= Eq(0, a)
```

Comment:
I just ported Chisel2's inference for backward compatibility and assumed enable signals are written according to the rules.